File: ITSA/2023-12-2/pE.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <cmath>

#define int long long
using namespace std;
bool cmp(pair<string,int> a,pair<string,int> b)
{
    return stoll(a.first) < stoll(b.first);
}
signed main()
{
    int n;
    cin >> n;
    vector<pair<string,int> > v(n);
    map<string,long double> a,b,c;
    map<string,long double> mp;
    for(int i=0;i<n;i++)
    {
        string s;
        cin >> s;
        a[s] = b[s] = c[s] = 0;
        cin >> a[s] >> b[s] >> c[s];
    }
    int q;
    cin >> q;
    for(int i=0;i<q;i++)
    {
        string s;
        int h;
        cin >> s >> h;
        if( mp.count(s) == 0 ) mp[s] = 0;
        mp[s] += h;
    }

    vector<pair<string, long double> > ans;

    for( auto i : a )
    {
        string s = i.first;
        long double total = a[s] * (long double)0.3 + b[s] * (long double)0.3 + c[s] * (long double)0.4;
        if( mp.count(s) != 0 && mp[s] > 6 ) total -= ( mp[s] - 6 );
        ans.push_back(make_pair(s,total));
    }

    sort(ans.begin(),ans.end(),cmp);
    for( auto i : ans )
    {
        cout << i.first << " ";
        int g = (int)i.second, t = i.second * 10;
        if( round(i.second) < 0 )
        {
            cout << 0 << "\n";
            continue;
        }
        cout << round(i.second) << "\n";
    }

    return 0;
}

File: ITSA/2023-12-2/pE.py
n = int(input())
v = []
a = {}
b = {}
c = {}
mp = {}
for i in range(n):
    w,x,y,z = input().split(" ")
    a[w] = b[w] = c[w] = 0
    a[w] = int(x)
    b[w] = int(y)
    c[w] = int(z)
q = int(input())
for i in range(q):
    s, h = input().split(" ")
    h = int(h)
    if s not in mp:
        mp[s] = 0
    mp[s] += h
ans = []
for s in a:
    total = a[s] * 0.3 + b[s] * 0.3 + c[s] * 0.4
    if s in mp and mp[s] > 6:
        total -= (mp[s] - 6)
    ans.append((s, total))
ans.sort(key=lambda x: int(x[0]))
for i in ans:
    print(i[0], end=" ")
    g = int(i[1])
    t = int(i[1] * 10)
    if round(i[1]) < 0:
        print(0)
        continue
    print(round(i[1]))

